File: pegasusX/apps/driver-app-android/app/src/main/java/com/pegasusx/driver/offline/DriverOfflineActionCatalog.kt
package com.pegasusx.driver.offline

import com.pegasusx.mobilekit.offline.OfflineEndpointCatalog
import com.pegasusx.mobilekit.offline.OfflineHttpSemantics

/**
 * Offline-eligible driver mutations and flush priority (lower = earlier).
 * Aligns with docs/big-platform-baseline/last-mile/4.1-offline-sync-protocol.md
 */
object DriverOfflineActionCatalog : OfflineEndpointCatalog {
    const val STATUS_PENDING = OfflineHttpSemantics.STATUS_PENDING
    const val STATUS_DEAD = OfflineHttpSemantics.STATUS_DEAD
    const val MAX_ATTEMPTS = OfflineHttpSemantics.MAX_ATTEMPTS_DEFAULT
    const val PROXIMITY_MAX_AGE_MS = 2 * 60 * 1000L

    const val ENDPOINT_PROXIMITY = "v1/delivery/proximity-unlock"
    const val ENDPOINT_SHOP_CLOSED = "v1/delivery/shop-closed"
    const val ENDPOINT_PARTIAL = "v1/delivery/partial-offload"
    const val ENDPOINT_DELIVER = "v1/order/deliver"
    const val ENDPOINT_COLLECT_CASH = "v1/order/collect-cash"
    const val ENDPOINT_CREDIT = "v1/delivery/credit-delivery"
    const val ENDPOINT_ARRIVE = "v1/delivery/arrive"
    const val ENDPOINT_OFFLOAD = "v1/order/confirm-offload"
    const val ENDPOINT_COMPLETE = "v1/order/complete"
    const val ENDPOINT_FISCAL_RETRY = "v1/order/{orderId}/fiscal/retry"
    const val ENDPOINT_SPLIT = "v1/delivery/split-payment"
    const val ENDPOINT_BYPASS_OFFLOAD = "v1/delivery/bypass-offload"
    const val ENDPOINT_PAYMENT_BYPASS = "v1/delivery/confirm-payment-bypass"
    const val ENDPOINT_DEPART = "v1/fleet/driver/depart"
    const val ENDPOINT_RETURN = "v1/fleet/driver/return-complete"
    const val ENDPOINT_CASH_RECON = "v1/driver/cash-reconciliations"
    const val ENDPOINT_ROUTE_REORDER = "v1/fleet/route/reorder"
    const val ENDPOINT_AVAILABILITY = "v1/driver/availability"

    override fun priorityFor(endpoint: String): Int = when (normalize(endpoint)) {
        ENDPOINT_PROXIMITY -> 10
        ENDPOINT_SHOP_CLOSED, ENDPOINT_PARTIAL, ENDPOINT_DELIVER -> 20
        ENDPOINT_COLLECT_CASH, ENDPOINT_CREDIT -> 30
        else -> 40
    }

    override fun isOfflineEligible(endpoint: String): Boolean {
        val ep = normalize(endpoint)
        return ep in setOf(
            ENDPOINT_PROXIMITY,
            ENDPOINT_SHOP_CLOSED,
            ENDPOINT_PARTIAL,
            ENDPOINT_DELIVER,
            ENDPOINT_COLLECT_CASH,
            ENDPOINT_CREDIT,
            ENDPOINT_ARRIVE,
            ENDPOINT_OFFLOAD,
            ENDPOINT_COMPLETE,
            ENDPOINT_SPLIT,
            ENDPOINT_BYPASS_OFFLOAD,
            ENDPOINT_PAYMENT_BYPASS,
            ENDPOINT_DEPART,
            ENDPOINT_RETURN,
            ENDPOINT_CASH_RECON,
            ENDPOINT_ROUTE_REORDER,
            ENDPOINT_AVAILABILITY,
        ) || ep.contains("/fiscal/retry")
    }

    override fun normalize(endpoint: String): String =
        OfflineHttpSemantics.normalizeEndpoint(endpoint)

    fun isRetryableHttp(code: Int): Boolean = OfflineHttpSemantics.isRetryableHttp(code)

    fun isSuccessHttp(code: Int): Boolean = OfflineHttpSemantics.isSuccessHttp(code)

    fun isNetworkEnqueueable(error: Throwable): Boolean {
        if (OfflineHttpSemantics.isNetworkEnqueueable(error)) return true
        return when (error) {
            is retrofit2.HttpException -> isRetryableHttp(error.code())
            else -> false
        }
    }
}

File: pegasusX/apps/driver-app-ios/driverappios/driverappios/Offline/DriverOfflineActionCatalog.swift
import Foundation
import PegasusKit

enum DriverOfflineActionStatus: String {
    case pending = "PENDING"
    case dead = "DEAD"
}

enum DriverOfflineActionCatalog {
    static let maxAttempts = OfflineHttpSemantics.maxAttemptsDefault
    static let proximityMaxAge: TimeInterval = 120

    static let proximity = "v1/delivery/proximity-unlock"
    static let shopClosed = "v1/delivery/shop-closed"
    static let partial = "v1/delivery/partial-offload"
    static let deliver = "v1/order/deliver"
    static let collectCash = "v1/order/collect-cash"
    static let credit = "v1/delivery/credit-delivery"
    static let arrive = "v1/delivery/arrive"
    static let offload = "v1/order/confirm-offload"
    static let complete = "v1/order/complete"
    static let split = "v1/delivery/split-payment"
    static let bypassOffload = "v1/delivery/bypass-offload"
    static let paymentBypass = "v1/delivery/confirm-payment-bypass"
    static let depart = "v1/fleet/driver/depart"
    static let returnComplete = "v1/fleet/driver/return-complete"
    static let cashRecon = "v1/driver/cash-reconciliations"
    static let routeReorder = "v1/fleet/route/reorder"
    static let availability = "v1/driver/availability"

    static func normalize(_ endpoint: String) -> String {
        OfflineHttpSemantics.normalizeEndpoint(endpoint)
    }

    static func priority(for endpoint: String) -> Int {
        switch normalize(endpoint) {
        case proximity: return 10
        case shopClosed, partial, deliver: return 20
        case collectCash, credit: return 30
        default: return 40
        }
    }

    static func isOfflineEligible(_ endpoint: String) -> Bool {
        let ep = normalize(endpoint)
        let known: Set<String> = [
            proximity, shopClosed, partial, deliver, collectCash, credit,
            arrive, offload, complete, split, bypassOffload, paymentBypass,
            depart, returnComplete, cashRecon, routeReorder, availability,
        ]
        return known.contains(ep) || ep.contains("/fiscal/retry")
    }

    static func isRetryableHTTP(_ code: Int) -> Bool {
        OfflineHttpSemantics.isRetryableHTTP(code)
    }

    static func isSuccessHTTP(_ code: Int) -> Bool {
        OfflineHttpSemantics.isSuccessHTTP(code)
    }

    /// True only for transport / retryable failures. Business 4xx (geofence, forbidden,
    /// invalid transition, etc.) must never enqueue — that caused P0-4 silent "success".
    static func isNetworkEnqueueable(_ error: Error) -> Bool {
        if let apiErr = error as? APIError {
            switch apiErr {
            case .networkError:
                return true
            case .httpError(let code):
                return isRetryableHTTP(code)
            case .problemDetail(let problem):
                if problem.retryable == true { return true }
                return isRetryableHTTP(problem.status)
            case .unauthorized, .forbidden, .decodingError, .invalidURL, .explainError:
                return false
            }
        }
        if let urlErr = error as? URLError {
            switch urlErr.code {
            case .notConnectedToInternet, .timedOut, .networkConnectionLost,
                 .cannotFindHost, .cannotConnectToHost, .dnsLookupFailed,
                 .internationalRoamingOff, .dataNotAllowed:
                return true
            default:
                return false
            }
        }
        let ns = error as NSError
        if ns.domain == NSURLErrorDomain {
            switch URLError.Code(rawValue: ns.code) {
            case .notConnectedToInternet, .timedOut, .networkConnectionLost,
                 .cannotFindHost, .cannotConnectToHost, .dnsLookupFailed,
                 .internationalRoamingOff, .dataNotAllowed:
                return true
            default:
                return false
            }
        }
        return false
    }

    static func nowIso() -> String {
        ISO8601DateFormatter().string(from: Date())
    }
}
